### ISSUE TITLE
Induction on dep bad evar instantiation

### ISSUE THREAD

**Issue by elefthei** [coq/coq#14674]
#### Description of the problem

> Error:
> Failed to instantiate evar: Anomaly "Uncaught exception Evarsolve.IllTypedInstance(_, _, _)."
> Please report at http://coq.inria.fr/bugs/.

```coq
From Coq Require Import Vectors.Vector.
From Coq Require Import Logic.Eqdep.
From Coq Require Import Program.Equality.

Import VectorNotations.

Section VPermutation.
  Context {A:Type}.

  Inductive Permutation: forall n, Vector.t A n -> Vector.t A n -> Prop :=
  | vperm_nil: Permutation 0 [] []
  | vperm_skip {n} x l l' : Permutation n l l' -> Permutation (S n) (x :: l) (x :: l')
  | vperm_swap {n} x y l : Permutation (S (S n)) (y :: x :: l) (x :: y :: l)
  | vperm_trans {n} l l' l'' :
      Permutation n l l' -> Permutation n l' l'' -> Permutation n l l''.

  Local Notation "a ~~ b" := (Permutation _ a b) (at level 41, left associativity).  

  (** Permutation lemmas *)
  Lemma permutation_singleton: forall a b,
      [a] ~~ [b] ->
      a = b.
  Proof.
    intros.
    remember ([a]) as m.
    induction H.
  Admitted.
End VPermutation.
```

#### Coq Version
coq 8.12

**Comment by Alizter** [comment]
This is the error message I get in 8.15+alpha:
```coq

Abstracting over the terms "n", "m" and "t" leads to a term
fun (n0 : nat) (m0 _ : Vector.t A n0) => m0 = [a] -> a = b
which is ill-typed.
Reason is: Illegal application: 
The term "@eq" of type "forall A : Type, A -> A -> Prop"
cannot be applied to the terms
 "Vector.t A n0" : "Type"
 "m0" : "Vector.t A n0"
 "[a]" : "Vector.t A 1"
The 3rd term has type "Vector.t A 1" which should be coercible to 
"Vector.t A n0".
```

**Comment by elefthei** [comment]
Confirmed, this is fixed in 8.13.0 as well. Feel free to close.
